### PR TITLE
Expand supported `RayError` exceptions

### PR DESCRIPTION
**Pull request by omus**
Creates Julia exceptions to mirror the Ray exceptions supported by Ray for Python.

One challenge with porting thing is that some exceptions are [MessagePack serialized containing pickle5 data](https://github.com/ray-project/ray/blob/a03efd9931128d387649dd48b0e4864b43d3bfb4/python/ray/_private/serialization.py#L287-L288). As we found out in #191 we can somewhat read the content without having to unpack it. I've opted to show users the `repr`'d formatted raw data for any exception we haven't seen before. This way users can complain about the bad errors and provide us with examples we can add to our test suites.

There is also some future work to be done around adding in call site information from Julia so we can deserialize that information on the driver and report it to the user. I've opted to leave that for now.

### CODE REVIEW

**Comment by omus** [comment]
Ugh, my merge wasn't so clean as it appeared

**Comment by omus** [comment]
Weird formatter failure:
```
Formatting /home/runner/work/Ray.jl/Ray.jl/build/bind_artifacts.jl
Invalid instruction at 0x7f4e3c60f297: 0x62, 0xf2, 0x7d, 0x48, 0x7c, 0xc0, 0x62, 0xf1, 0x7d, 0x48, 0xfe, 0x0d, 0x59, 0x1e, 0xff

[1909] signal (4.2): Illegal instruction
in expression starting at /home/runner/work/_temp/1d519080-2ee3-48ad-9e23-52a55bb422f8:5
dotop1 at /home/runner/.julia/packages/Tokenize/d4Pxs/src/utilities.jl:45 [inlined]
lex_dot at /home/runner/.julia/packages/Tokenize/d4Pxs/src/lexer.jl:904
next_token at /home/runner/.julia/packages/Tokenize/d4Pxs/src/lexer.jl:380
next_token at /home/runner/.julia/packages/Tokenize/d4Pxs/src/lexer.jl:315 [inlined]
next at /home/runner/.julia/packages/CSTParser/VcYj6/src/lexer.jl:97
#parse_expression#19 at /home/runner/.julia/packages/CSTParser/VcYj6/src/CSTParser.jl:49
parse_expression at /home/runner/.julia/packages/CSTParser/VcYj6/src/CSTParser.jl:41
parse_expression at /home/runner/.julia/packages/CSTParser/VcYj6/src/CSTParser.jl:41
parse_doc at /home/runner/.julia/packages/CSTParser/VcYj6/src/CSTParser.jl:220
parse at /home/runner/.julia/packages/CSTParser/VcYj6/src/CSTParser.jl:248
format_text at /home/runner/.julia/packages/JuliaFormatter/ur5KV/src/JuliaFormatter.jl:239
#format_text#251 at /home/runner/.julia/packages/JuliaFormatter/ur5KV/src/JuliaFormatter.jl:215
format_text at /home/runner/.julia/
...
```

**Comment by omus** [comment]
3rd run of the formatting workflow worked

File: src/exceptions.jl [hunk 1]
```diff
@@ -1,16 +1,81 @@
...
 # https://github.com/ray-project/ray/blob/ray-2.5.1/python/ray/exceptions.py
 
+struct ObjectContext
+    object_ref_hex::String
+    owner_address::ray_jll.Address
+    call_site::String
+end
+
+function ObjectContext(obj_ref::ObjectRef)
+    return ObjectContext(hex_identifier(obj_ref), get_owner_address(obj_ref), "")
```
Comment (kleinschmidt) [inline src/exceptions.jl:11]: is this really not covered by tests or is it an inlining failure?

File: src/exceptions.jl [hunk 2]
```diff
@@ -28,6 +93,295 @@ end
...
+
+function Base.showerror(io::IO, ex::ObjectReconstructionFailedMaxAttemptsExceededError)
+    print(io, "$ObjectReconstructionFailedMaxAttemptsExceededError: ")
+    print_object_lost(io, ex.object_context)
+
+    # TODO: Update this message with more details on how to set `max_retries` once
+    # implemented: https://github.com/ray-project/ray/blob/ray-2.5.1/python/ray/exceptions.py#L593
+    print(io, m"""
+              The object cannot be reconstructed because the maximum number of task retries
+              has been exceeded. To prevent this error, set `max_retries` (default 3).
```
Comment (kleinschmidt) [inline src/exceptions.jl:291]: I'm not sure this is accurate; with #215 I think we default to 0? https://github.com/beacon-biosignals/Ray.jl/blob/6e11dec0bcdbbfbf5286a4929d22280334dc6307/src/runtime.jl#L144

File: src/exceptions.jl [hunk 3]
```diff
@@ -28,6 +93,295 @@ end
...
+    return nothing
+end
+
+"""
+    ObjectFreedError <: RayError
+
+Indicates that an object was manually freed by the application.
+
+Currently should never happen as Ray.jl doesn't currently implement `free` like
+[Ray for Python does](https://github.com/ray-project/ray/blob/ray-2.5.1/python/ray/_private/internal_api.py#L170).
```
Comment (kleinschmidt) [inline src/exceptions.jl:195]: ah sonds like something we _should_ implement though (woudl help with some of the "disk full" errors we've seen during benchmarking big cluster jobs...)

**Review by kleinschmidt — APPROVED** [review]
a few comments.  I didn't review this very carefully but let's just merge and wrap this bit up.

File: src/exceptions.jl (hunk 3 above)
Comment (kleinschmidt) [reply src/exceptions.jl:195]: https://github.com/beacon-biosignals/Ray.jl/issues/230

File: src/exceptions.jl (hunk 1 above)
Comment (omus) [reply src/exceptions.jl:11]: It's not covered by tests. This is the only part of the this code that requires Ray.jl to be connected to the raylet. As the majority of the code doesn't require it I opted to test this source file without being connected to the raylet. As a follow up I think I'd make a convert method in `object_ref.jl` and test this portion that way.

I'll make an issue

File: src/exceptions.jl (hunk 2 above)
Comment (omus) [reply src/exceptions.jl:291]: Good catch. I wrote this code I believe before we implemented retries

File: src/exceptions.jl (hunk 1 above)
Comment (omus) [reply src/exceptions.jl:11]: https://github.com/beacon-biosignals/Ray.jl/issues/231

**Comment by omus** [comment]
Alright, finally going to merge this PR. I need to bump the version yet but I'll let CI finish first.

**Pull request merged**